internal/addrs/move_endpoint_module_test.go

```diff
@@ -1074,3 +1074,179 @@ func TestAbsResourceMoveDestination(t *testing.T) {
+				Call:   ModuleCall{Name: "bar"},
+			},
+			Other:        mustParseModuleInstanceStr("module.foo[2]"),
+			CanChainFrom: false,
+			NestedWithin: false,
+		},
+
+		{
+			Endpoint:     mustParseModuleInstanceStr("module.foo[2]"),
+			Other:        mustParseModuleInstanceStr("module.foo[2]"),
+			CanChainFrom: true,
+			NestedWithin: false,
+		},
+
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			Other:        mustParseModuleInstanceStr("module.foo[2]"),
+			CanChainFrom: false,
+			NestedWithin: true,
+		},
+
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].module.bar.resource.baz"),
+			Other:        mustParseModuleInstanceStr("module.foo[2]"),
+			CanChainFrom: false,
+			NestedWithin: true,
+		},
+
+		{
+			Endpoint: AbsModuleCall{
+				Module: mustParseModuleInstanceStr("module.foo[2]"),
+				Call:   ModuleCall{Name: "bar"},
+			},
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			CanChainFrom: false,
+			NestedWithin: false,
+		},
+
+		{
+			Endpoint:     mustParseModuleInstanceStr("module.foo[2]"),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			CanChainFrom: false,
+			NestedWithin: false,
+		},
+
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			CanChainFrom: true,
+			NestedWithin: false,
+		},
+
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz[2]").ContainingResource(),
+			CanChainFrom: false,
+			NestedWithin: true,
+		},
+
+		{
+			Endpoint: AbsModuleCall{
+				Module: mustParseModuleInstanceStr("module.foo[2]"),
+				Call:   ModuleCall{Name: "bar"},
+			},
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			CanChainFrom: false,
+		},
+		{
+			Endpoint:     mustParseModuleInstanceStr("module.foo[2]"),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			CanChainFrom: false,
+		},
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].resource.baz").ContainingResource(),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			CanChainFrom: false,
+		},
+		{
+			Endpoint:     mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			Other:        mustParseAbsResourceInstanceStr("module.foo[2].resource.baz"),
+			CanChainFrom: true,
+		},
+	}
+
+	for i, test := range tests {
+		t.Run(fmt.Sprintf("[%02d]%s.CanChainFrom(%s)", i, test.Endpoint, test.Other),
+			func(t *testing.T) {
+				endpoint := &MoveEndpointInModule{
+					relSubject: test.Endpoint,
+				}
+
+				other := &MoveEndpointInModule{
+					relSubject: test.Other,
+				}
```

Comment:
This doesn't seem to have any test coverage for the situations where `endpoint.module` or `other.module` are non-empty -- that is, when a `moved` block appears in a nested module and thus needs to match only within the instances of that module. It looks like the function implementations themselves don't handle that situation either.

It's within those situations that the more tricky cases arise. For example:

* In a situation where the root module declares something like this:

    ```hcl
    moved {
      from = module.foo
      to   = module.bar
    }
    ```

    ...and then inside the "bar" module there's also a declaration like this:

    ```hcl
    moved {
      from = aws_instance.foo
      to   = aws_instance.bar
    }
    ```

    In that case, the second move is "nested within" the first move, because we need to ensure a deterministic ordering for processing them. (In that case, a `module.foo.aws_instance.foo` in the state ought to be treated as `module.bar.aws_instance.bar` once we've processed both of these move statements.

* In a situation where the root module declares something like this:

    ```hcl
    moved {
      from = aws_instance.foo
      to   = module.foo.aws_instance.foo
    }
    ```

    ...and then the "foo" module has a declaration like this:

    ```hcl
    moved {
      from = aws_instance.foo
      to   = aws_instance.bar
    }
    ```

    These two statements must follow the chaining rule, so that an `aws_instance.foo` in the state can end up being treated as `module.foo.aws_instance.bar` once we've processed both statements.

Reply:
Ah sorry, I didn't elaborate on that in the PR comment. This is not complete, and only covers the simple single instance cases. Chaining through modules was not yet working (there ware some issues in existing code which I can't recall at the moment), and the test fixtures and resulting data structures are going to require some refactoring anyway, I was leaving that for the next PR.

If you think this is a satisfactory incremental PR, we can commit this to avoid it growing too large. Otherwise I can continue work on top of this PR to make it more feature complete.

Reply:
Merging this checkpoint seems fine to me! Just wanted to make sure I was understanding the situation correctly. I guess it could be helpful to add some test cases (disabled, for now) for situations like what I wrote out in this comment so that we've got some cases to work towards, but doesn't really matter if we do that in this PR or in the next one because they'd need to be disabled in this one anyway.